### ISSUE TITLE
Totality checker happy with single decreasing recursive call

### ISSUE THREAD

**Issue by buzden** [idris-lang/Idris2#2448]
# Steps to Reproduce

It was originally discovered on a some complex data structure but works for simpler ones.

Consider a named instance that uses the interface's operation inside:

```idris
%default total

[W] Eq Nat where
  x == y = x == y
```

This does not typecheck showing that `==` function from inside the implementation considers this named instance to be found in `auto` search (one can check that if we remember the result of `%search` before the definition and use it in the implementation, everything typechecks, so behaviour is consistent).

Now, add a pattern matching clause before:

```idris
%default total

[W] Eq Nat where
  S x == S y = x == y
  x   == y   = x == y
```

# Expected Behavior

Fails to typecheck either.

# Observed Behavior

Typechecks. Being run `(0 == 0) @{W}`, results to an infinite loop.

**Comment by gallais** [comment]
Without the unrelated interfaces and without reusing an existing type which already has an `Eq`
implementation to avoid confusion wrt which implementation is picked on the RHS:

```
%default total                                                                                                                  
 
data Nat2 = Z | S Nat2
 
eqNat2 : Nat2 -> Nat2 -> Bool
eqNat2 (S x) (S y) = eqNat2 x y
eqNat2 x y = eqNat2 x y
```

**Comment by gallais** [comment]
This was solved by #2877
